libnetdata/dictionary/dictionary.c

```diff
@@ -253,78 +582,362 @@ int dictionary_del(DICTIONARY *dict, const char *name) {
     return ret;
 }
 
+// ----------------------------------------------------------------------------
+// traversal with loop
+
+void *dictionary_foreach_start(DICTFE *dfe, DICTIONARY *dict) {
+    if(unlikely(!dfe || !dict)) return NULL;
+
+    dfe->dict = dict;
+    dfe->started_ut = now_realtime_usec();
+
+    dictionary_read_lock(dict);
+
+    NAME_VALUE *nv = dict->first_item;
+    dfe->last_position_index = (void *)nv;
+
+    if(likely(nv)) {
+        dfe->name = nv->name;
+        dfe->value = (void *)nv->value;
+
+        // TODO: increment reference counter on nv;
+    }
+    else {
+        dfe->name = NULL;
+        dfe->value = NULL;
+    }
+
+    return dfe->value;
+}
+
+void *dictionary_foreach_next(DICTFE *dfe) {
+    if(unlikely(!dfe || !dfe->dict)) return NULL;
+
+    NAME_VALUE *nv = (NAME_VALUE *)dfe->last_position_index;
+    if(likely(nv)) {
+        // TODO: decrement reference counter on nv
+
+        // get the next item
+        nv = nv->next;
+    }
+    dfe->last_position_index = (void *)nv;
+
+    if(likely(nv)) {
+        dfe->name = nv->name;
+        dfe->value = (void *)nv->value;
+
+        // TODO: increment reference counter on nv->next;
+    }
+    else {
+        dfe->name = NULL;
+        dfe->value = NULL;
+    }
+
+    return dfe->value;
+}
+
+usec_t dictionary_foreach_done(DICTFE *dfe) {
+    if(unlikely(!dfe || !dfe->dict)) return 0;
+
+    NAME_VALUE *nv = (NAME_VALUE *)dfe->last_position_index;
+    if(nv) {
+        // TODO: decrement reference counter on nv
+        ;
+    }
+
+    dictionary_unlock((DICTIONARY *)dfe->dict);
+    dfe->dict = NULL;
+    dfe->last_position_index = NULL;
+    dfe->name = NULL;
+    dfe->value = NULL;
+
+    usec_t usec = now_realtime_usec() - dfe->started_ut;
```

Comment:
Here you are calculating a value that is never used

Reply:
it is returned 2 lines below

Reply:
Sorry by my fault, I was reviewing too many consecutive lines, and I did not observe this :man_facepalming: